nixos: Data module imports

zimbatm opened the pull request:
###### Motivation for this change

This extends the capabilities of NixOS modules to be able to import JSON or TOML configurations :)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

zimbatm committed:
lib: add importTOML

Complements the `lib.importJSON`. `builtins.readTOML` has been
introduced in Nix 2.1.

Infinisil commented:
I don't like the idea of adding such implicit behavior. How about an explicit usage like this instead:
```nix
{ lib, ... }: {
  imports = [
    (lib.modules.fromJSON ./path/to.json)
  ];
}
```

Which would be implemented like
```nix
{
  fromJSON = file: {
    _file = file;
    config = importJSON file;
  };
}
```

zimbatm committed:
lib: allow to import JSON and TOML files

The vision here is that configuration tools can generate .json or .toml
files, which can be plugged into an existing configuration.

Eg:

    { lib, ... }:
    {
      imports = [
        (lib.modules.importJSON ./hardware-configuration.json)
      ];
    }

zimbatm commented:
Great idea @Infinisil . Updated!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/having-a-go-at-using-toml-to-configure-nixos/13897/1
